numba/parfor.py

```diff
@@ -3837,32 +3844,50 @@ def process_assign(stmt):
                 for s in stmt.init_block.body:
                     process_assign(s)
                 pblocks = stmt.loop_body.copy()
-                push_call_vars(pblocks, saved_globals, saved_getattrs, nested=True)
+                push_call_vars(pblocks, saved_globals, saved_getattrs, typemap, nested=True)
                 new_body.append(stmt)
                 continue
             else:
                 process_assign(stmt)
             for v in stmt.list_vars():
                 new_body += _get_saved_call_nodes(v.name, saved_globals,
-                                                  saved_getattrs, block_defs)
+                                                  saved_getattrs, block_defs, rename_dict)
             new_body.append(stmt)
         block.body = new_body
+        # If there is anything to rename then apply the renaming here.
+        if len(rename_dict) > 0:
+            # Fix-up the typing for the renamed vars.
+            for k,v in rename_dict.items():
+                typemap[v] = typemap[k]
+            # This is only to call replace_var_names which takes a dict.
+            temp_blocks = {0:block}
+            replace_var_names(temp_blocks, rename_dict)
 
     return
 
 
-def _get_saved_call_nodes(fname, saved_globals, saved_getattrs, block_defs):
+def _get_saved_call_nodes(fname, saved_globals, saved_getattrs, block_defs, rename_dict):
     nodes = []
     while (fname not in block_defs and (fname in saved_globals
                                         or fname in saved_getattrs)):
         if fname in saved_globals:
-            nodes.append(saved_globals[fname])
-            block_defs.add(saved_globals[fname].target.name)
+            assert(isinstance(saved_globals[fname], ir.Assign))
+            saved_global = saved_globals[fname]
+            renamed_var = ir.Var(saved_global.target.scope, mk_unique_var("$push_global_to_block"), saved_global.target.loc)
+            renamed_assign = ir.Assign(copy.deepcopy(saved_global.value), renamed_var, saved_global.loc)
+            nodes.append(renamed_assign)
+            block_defs.add(saved_global.target.name)
+            rename_dict[saved_global.target.name] = renamed_assign.target.name
             fname = '_PA_DONE'
         elif fname in saved_getattrs:
+            assert(isinstance(saved_getattrs[fname], ir.Assign))
             up_name = saved_getattrs[fname].value.value.name
-            nodes.append(saved_getattrs[fname])
-            block_defs.add(saved_getattrs[fname].target.name)
+            saved_getattr = saved_getattrs[fname]
+            renamed_var = ir.Var(saved_getattr.target.scope, mk_unique_var("$push_getattr_to_block"), saved_getattr.target.loc)
+            renamed_assign = ir.Assign(copy.deepcopy(saved_getattr.value), renamed_var, saved_getattr.loc)
+            nodes.append(renamed_assign)
+            block_defs.add(saved_getattr.target.name)
+            rename_dict[saved_getattr.target.name] = renamed_assign.target.name
```

Comment:
The code on both sides of the branch looks pretty much the same, could it be a local function with `saved_getattr, saved_getattrs, fname` and `saved_global, saved_globals, fname` or similar as args?

Reply:
Done.